### PR TITLE
feat(training): add length-aware batching and isolated sequence packing

### PR DESCRIPTION
**Pull request by Juzezhang**
## Summary

Add opt-in length-aware microbatching for local LeRobot latent training with the DualExpert Video-Then-Action (VTA) and Joint programs. This enables multiple variable-length trajectories per microbatch while preserving each sample's conditioning, visibility rules, supervision boundaries, and loss normalization.

Existing recipes continue to use `data.batching.mode: strict`. No existing training job or checkpoint is modified.

### Implementation

- **Length bucketing (`bucket`):** sort bounded pools of the existing sampler's indices by metadata-only length hints, then dynamically pad. Retained replacement draws and their multiplicities are preserved; incomplete training tails are removed before sorting.
- **Dynamic padding (`padded`):** retain original tensor extents and masks, exclude batch-added padding from conditioning and supervised losses, and keep fixed-size training microbatches.
- **Isolated sequence packing (`packed`):** combine real video/action tokens into one heavy transformer execution per layer. Both self-attention and text/proprioceptive cross-attention enforce sample boundaries; positions and within-sample attention semantics remain intact. CUDA uses FlexAttention block masks.
- **Loss and validation accounting:** reuse the existing decoder for each unpadded sample, average per-sample losses equally, and aggregate variable-batch validation metrics by original sample count, including incomplete tails.
- Add professional user documentation, configuration examples, public documentation navigation, and forward/backward isolation and parity coverage.

## Change Type

- [x] Runtime modernization
- [x] Opt-in behavior change
- [ ] Legacy removal

## Compatibility And Scope

- Existing YAMLs, CLI entrypoints, checkpoint parameter keys, and the default `strict` path remain supported.
- New configuration is grouped under `data.batching`; no new dependencies are introduced.
- Supported scope is `lerobot_v2_latent_local` + `uniform_segment`, the latent batch adapter, DualExpert VTA/Joint, and `sample_loss_weight_mode: none`. Unsupported sources, routing, and programs fail explicitly.
- Training keeps a fixed sample count per rank. This is not a dynamic token-budget batch sampler. Shared chunk/window geometry intentionally correlates samples within a microbatch; changing batching recipes is not a claim of step-for-step stochastic replay.
- The serialized mixed-video artifact golden was updated only for the additive default `batching` stanza. Removing that stanza reproduces the previous artifact fingerprint exactly.

## Validation

Validated locally with the repository's frozen environment: Python 3.12.13, PyTorch 2.11.0, Diffusers 0.37.1, and Transformers 5.10.4 on macOS.

- Focused data/config/runtime/model/checkpoint/architecture/documentation regression: **1,005 passed, 1 skipped, 2 deselected**.
- Updated mixed-video serialization golden: **1 passed**.
- Full required CPU selection: **2,638 passed, 26 failed, 8 skipped, 11 deselected**. The full suite is **not green on this macOS host**: all 26 remaining failures reproduce on unmodified `main` (`88592eb`) in the same environment:
  - 24 existing launcher tests encounter `${value,,}` with the system's Bash 3.2.
  - One mixed-video timestamp-cache assertion reproduces identically; codec sensitivity is suspected but not conclusively isolated.
  - One seeding test unconditionally asserts a cuDNN flag although cuDNN is unavailable on this platform.
- Lockfile verification, no-Torch static sanity, repository-wide Pyflakes, and `git diff --check`: **passed**.
- Public documentation staging and strict MkDocs build: **passed**, with no private-context leaks or broken references.
- Wheel build and installed-distribution checks outside the source checkout: **passed**.

Representative commands:

```bash
uv sync --frozen --group dev --extra full
TMPDIR=/private/tmp CUDA_VISIBLE_DEVICES="" OMP_NUM_THREADS=1 \
  uv run --frozen --extra full --group dev pytest --strict-markers -q \
  --tb=no -m "not (gpu or sim or data or slow)"
uv lock --check
uv run python -m pyflakes src scripts tests baselines
uv run python scripts/build_docs_site.py --output .docs_site
uv run mkdocs build --strict
git diff --check
```

## Resource Limits And Follow-Up Validation

The new CUDA kernel tests are marked `gpu` and require `OPEN_WAM_RUN_GPU_SANITY=1`; they were **not executed** on this CPU-only host. They cover compiled self/cross FlexAttention, activation-checkpointed backward parity, and cross-sample isolation.

Representative CUDA/FSDP, precision, long-sequence memory, compilation overhead, and throughput validation remain required before adopting a new mode for a long training run. This PR makes no measured speedup or memory-saving claim. Private datasets, checkpoints, machine paths, and deployment records are not included.

### CODE REVIEW

**Comment by DaivdYuan** [comment]
This contribution has been imported into the private staging workflow as DaivdYuan/OpenWAM-staging#64, with Juze Zhang's authorship and the source commit provenance preserved. The source tree was verified identical before staging-only test hardening. Please avoid parallel edits here while staging review runs; after staging merges, the promotion workflow will open the canonical public replacement PR and this PR can be closed as superseded.

**Comment by DaivdYuan** [comment]
Superseded by #7
